### PR TITLE
Channel to object FIFO: 1-to-1 L2 support

### PR DESCRIPTION
**Pull request by abisca**
This PR introduces 1-to-1 L2 support in the channel to object FIFO lowering. Specifically, it leverages the existing MemTile allocator to allocate L2 buffers to MemTiles. That information is stored in a map and forwarded to the object FIFO lowering pass which then uses it to generate objectFifos with MemTiles and LinkOps, where applicable (i.e., if the L2 result from a get from one channel is put into another channel). The allocated L2 buffers are erased as object FIFOs allocate their own.

Two examples are part of this PR: one with channel-to-objectFifo lowering of 1-to-1 channels with L3-L2-L1 data movement, and the other with channels using broadcast and L3-L2-L1 data movement.

Support for join and distribute patterns will be part of a subsequent PR.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,6 +17,7 @@ _x
 vck5000_aie_base_graph_hw.ltx
 vck5000_aie_base_graph_hw.xsa.link_summary
 Work
+sandbox
```
Comment (erwei-xilinx) [inline .gitignore:20]: Would be great to remove this, as virtualenv name doesn't have to be called "sandbox".

Comment (abisca) [reply .gitignore:20]: Ah yes, good point. I will remove it!

**Review by jgmelber — APPROVED** [review]
LGTM

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir [hunk 2]
```diff
@@ -0,0 +1,74 @@
+//===- air_channel_to_objectfifo_L1toL2.mlir --------------------*- MLIR -*-===//
+//
+// Copyright (C) 2022, Advanced Micro Devices, Inc.
```
Comment (fifield) [inline mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir:3]: 2024

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// SPDX-License-Identifier: MIT
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: air-opt %s -air-place-herds='num-rows=2 num-cols=2 row-anchor=3 col-anchor=5' --air-to-aie='emit-while-loop=false row-offset=3 col-offset=5 use-objectfifo=true device=xcve2802' --canonicalize | FileCheck %s
```
Comment (fifield) [inline mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir:8]: Can this command line be simplified to only `air-to-aie`?  `emit-while-loop=false` is already the default

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir [hunk 4]
```diff
@@ -0,0 +1,83 @@
+//===- air_channel_to_objectfifo_L2_broadcast.mlir --------------------*- MLIR -*-===//
+//
+// Copyright (C) 2022, Advanced Micro Devices, Inc.
```
Comment (fifield) [inline mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir:3]: 2024

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// SPDX-License-Identifier: MIT
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: air-opt %s -air-place-herds='num-rows=2 num-cols=2 row-anchor=3 col-anchor=5' --air-to-aie='emit-while-loop=false row-offset=3 col-offset=5 use-objectfifo=true device=xcve2802' --canonicalize | FileCheck %s
```
Comment (fifield) [inline mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir:8]: Can this command line be simplified to only `air-to-aie`?

**Review by fifield — APPROVED** [review]

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir (hunk 4 above)
Comment (AndraBisca) [reply mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L2_broadcast.mlir:3]: Noted, thank you Jeff!

File: mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir (hunk 3 above)
Comment (AndraBisca) [reply mlir/test/Conversion/AIRToAIE/air_channel_to_objectfifo_L1toL2.mlir:8]: Good point. I'll take a look and simplify.

**Pull request merged**